mlir_graphblas/engine.py

```diff
@@ -261,65 +246,11 @@ def input_tensor_to_ctypes(
     tensor_type: mlir.astnodes.RankedTensorType,
 ) -> Tuple[list, Callable]:
 
-    input_c_types = []
-    np_type, pointer_type = convert_mlir_atomic_type(
-        tensor_type.element_type, return_pointer_type=True
-    )
-    input_c_types.append(pointer_type)  # allocated pointer (for free())
-    input_c_types.append(pointer_type)  # base pointer
-    input_c_types.append(ctypes.c_int64)  # offset from base
-    for _ in range(2 * len(tensor_type.dimensions)):  # dim sizes and strides
-        input_c_types.append(ctypes.c_int64)
-
-    dimensions = [dim.value for dim in tensor_type.dimensions]
-
-    def encoder(arg: np.ndarray) -> list:
-        if not isinstance(arg, np.ndarray):
-            raise TypeError(
-                f"{repr(arg)} is expected to be an instance of {np.ndarray.__qualname__}"
-            )
-        if not arg.dtype == np_type:
-            raise TypeError(f"{repr(arg)} is expected to have dtype {np_type}")
-        if not len(dimensions) == len(arg.shape):
-            raise ValueError(
-                f"{repr(arg)} is expected to have rank {len(dimensions)} but has rank {len(arg.shape)}."
-            )
-
-        encoded_args = [arg, arg, 0]
-
-        for dim_index, dim_size in enumerate(arg.shape):
-            expected_dim_size = dimensions[dim_index]
-            if (
-                expected_dim_size is not None
-                and arg.shape[dim_index] != expected_dim_size
-            ):
-                raise ValueError(
-                    f"{repr(arg)} is expected to have size {expected_dim_size} in the {dim_index}th dimension but has size {arg.shape[dim_index]}."
-                )
-            encoded_args.append(arg.shape[dim_index])
-
-        for dimension_index in range(len(arg.shape)):
-            stride = arg.strides[dimension_index] // arg.itemsize
-            encoded_args.append(stride)
-
-        return encoded_args
-
-    return input_c_types, encoder
-
-
-def input_llvm_pointer_to_ctypes(
-    mlir_type: _DIALECT_TYPES["llvm"]["LLVMPtr"],
-) -> Tuple[list, Callable]:
-    if (
-        isinstance(mlir_type.type, mlir.astnodes.IntegerType)
-        and int(mlir_type.type.width) == 8
-    ):
-        # TODO we blindly assume that an i8 pointer points to a sparse tensor
-        # since MLIR's sparse tensor support is currently up-in-the-air and this
-        # is how they currently handle sparse tensors
-        type_string = mlir_type.type.dump()
-        ctypes_type = LLVM_DIALECT_TYPE_STRING_TO_CTYPES_POINTER_TYPE[type_string]
-        ctypes_input_types = [ctypes_type]
+    if tensor_type.encoding is not None:
+        # We assume anything with an encoding must be a sparse tensor
+        # After lowering, this will become !llvm.ptr<i8>
+        ctypes_type = ctypes.POINTER(ctypes.c_int8)
+        input_c_types = [ctypes_type]
 
         def encoder(arg: MLIRSparseTensor) -> list:
```

Comment:
This is more of a "would be nice" suggestion than a requirement, but I think we can add a more specific type hint here of `def encoder(arg: MLIRSparseTensor) -> List[ctypes_type]:`.

It's not super important, so feel free to ignore this suggestion.